pybamm/expression_tree/concatenations.py

```diff
@@ -232,3 +248,57 @@ def simplify(self):
         new_node.domain = []
 
         return pybamm.simplify_if_constant(new_node)
+
+
+class SparseStack(pybamm.Symbol):
+    """A node in the expression tree representing a concatenation of sparse
+    matrices. As with NumpyConcatenation, we *don't* care about domains.
+    The class :class:`pybamm.DomainConcatenation`, which *is* careful about
+    domains and uses broadcasting where appropriate, should be used whenever
+    possible instead.
+
+    **Extends**: :class:`pybamm.Symbol`
+
+    Parameters
+    ----------
+    children : iterable of :class:`pybamm.Symbol`
+        The equations to concatenate
+
+    """
+
+    def __init__(self, *children):
+        super().__init__("model concatenation", children, domain=[])
+
+    def evaluate(self, t=None, y=None):
+        """ See :meth:`pybamm.Symbol.evaluate()`. """
+        if len(self.children) == 0:
+            return np.array([])
+        else:
+            return vstack([self.evaluate_child(child, t, y) for child in self.children])
+
+    def evaluate_child(self, child, t, y):
+        """ Evaluates the child and, if the results is dense, converts the
+        result as a sparse matrix.
+        """
+        # NOTE: I think this probably a very hacky way of doing this, but need
+        # some way of concatenating sparse matrices with dense matrices that
+        # come from evaluate. The other way would be to make all the matrices
+        # dense and then do a NumpyConcatenation, but I fear this will be bad
+        # when the Jacobian is very large
```

Comment:
new treatment of functions should solve this?

Reply:
Yep, this can be resolved once we replace `node_to_edge` etc. with matrix multiplication